src/System.Text.Json/tests/JsonDocumentTests.cs

```diff
@@ -3712,6 +3712,43 @@ private static void AssertContents(string expectedValue, ArrayBufferWriter<byte>
 #endif
                     ));
         }
+
+        [Fact]
+        // Issue #39930
+        public static void Disposing_On_Multiple_Threads_Should_Not_Return_To_ArrayPool_Twice()
+        {
+            Action<object> disposeAction = (object document) => ((JsonDocument)document).Dispose();
+
+            // Create a bunch of parallel tasks that call Dispose several times on the same object.
+            Task[] tasks = new Task[100];
+            int count = 0;
+            for (int j = 0; j < 10; j++)
+            {
+                JsonDocument document = JsonDocument.Parse("123" + j);
+                for (int i = 0; i < 10; i++)
+                {
+                    tasks[count] = new Task(disposeAction, document);
+                    tasks[count].Start();
+
+                    count++;
+                }
+            }
+
+            Task.WaitAll(tasks);
+
+            // When ArrayPool gets corrupted, the Rent method might pull an already rented array, which is incorrect.
+            // So we will pull as many arrays as calls to JsonElement.Dispose and check they are unique.
+            HashSet<IntPtr> uniqueAddresses = new HashSet<IntPtr>();
+            while (count > 0)
+            {
+                byte[] arr = ArrayPool<byte>.Shared.Rent(4);
```

Comment:
nit: Consider adding a comment explaining why 4 is being used (or maybe we should explicitly ask for 16 here). It's in relation to the array size being rented by `JsonDocument.Parse(...)` given the input is a 4-character JSON (i.e. "123X").

Reply:
that is indeed the reason why I ask for 4.
Also, I am not sure if the bucket sizes are the same in all platforms. that's why I would consider in keeping it as 4.

Reply:
> that is indeed the reason why I ask for 4.

Add that comment in source.